### PR TITLE
Add cloning functional tests.

### PR DESCRIPTION
**Pull request by awels**
Signed-off-by: Alexander Wels <awels@redhat.com>

**What this PR does / why we need it**:
Add functional tests for host assisted cloning.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by awels** [comment]
ci test please

**Comment by awels** [comment]
ci test please

**Comment by awels** [comment]
ci test please

**Comment by awels** [comment]
ci test please

File: tests/cloner_test.go [hunk 1]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+			Expect(err).ToNot(HaveOccurred())
+		}
+		if targetDv3 != nil {
+			By("Cleaning up target3 DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv3.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+	})
+
+	It("Should allow multiple clones from a single source datavolume", func() {
```
Comment (kgoldbla) [inline tests/cloner_test.go]: Please add the Polarion ID as follows:
It("[rfe_id:CNV-1277][case_id:CNV-1891][crit:High][vendor:cnv-qe@redhat.com][level:component] Should allow multiple clones from a single source datavolume", func() {

File: tests/cloner_test.go [hunk 2]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+		_, err = utils.CreateDataVolumeFromDefinition(f.CdiClient, f.Namespace.Name, targetDv3)
+		Expect(err).ToNot(HaveOccurred())
+		By("Waiting for clone to be completed")
+		err = utils.WaitForDataVolumePhase(f.CdiClient, f.Namespace.Name, cdiv1.Succeeded, targetDv3.Name)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(f.VerifyTargetPVCContentMD5(f.Namespace, utils.PersistentVolumeClaimFromDataVolume(targetDv3), matchFile, md5sum[:32])).To(BeTrue())
+	})
+})
+
+var _ = Describe("Validate Data Volume clone to smaller size", func() {
```
Comment (kgoldbla) [inline tests/cloner_test.go]: Please add the Polarion ID as follows:
var _ = Describe("[rfe_id:CNV-1126][case_id:CNV-1896][crit:High][vendor:cnv-qe@redhat.com][level:component] Validate Data Volume clone to smaller size", func() {

File: tests/cloner_test.go [hunk 3]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+		Expect(err).ToNot(HaveOccurred())
+		By("Waiting for clone to be completed")
+		err = utils.WaitForDataVolumePhase(f.CdiClient, f.Namespace.Name, cdiv1.Succeeded, targetDv.Name)
+		Expect(err).ToNot(HaveOccurred())
+		matchFile := filepath.Join(testBaseDir, "disk.img")
+		Expect(f.VerifyTargetPVCContentMD5(f.Namespace, utils.PersistentVolumeClaimFromDataVolume(targetDv), matchFile, md5sum[:32])).To(BeTrue())
+	})
+})
+
+var _ = Describe("Validate Data Volume should clone multiple clones in parallel", func() {
```
Comment (kgoldbla) [inline tests/cloner_test.go]: Please add the Polarion ID as follows:
[rfe_id:CNV-1277][case_id:CNV-1899][crit:High][vendor:cnv-qe@redhat.com][level:component]

File: tests/cloner_test.go [hunk 4]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+			Expect(err).ToNot(HaveOccurred())
+		}
+		if targetDv3 != nil {
+			By("Cleaning up target3 DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv3.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+	})
+
+	It("Should allow multiple cloning operations in parallel", func() {
```
Comment (kgoldbla) [inline tests/cloner_test.go]: Please add the Polarion ID as follows:
[rfe_id:CNV-1277][case_id:CNV-1899][crit:High][vendor:cnv-qe@redhat.com][level:component]

File: tests/cloner_test.go [hunk 5]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+			Expect(err).ToNot(HaveOccurred())
+		}
+		if targetDv != nil {
+			By("Cleaning up target DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+	})
+
+	It("Should not allow cloning into a smaller sized data volume", func() {
```
Comment (kgoldbla) [inline tests/cloner_test.go]: Please add the Polarion ID as follows:
It("[rfe_id:CNV-1126][case_id:CNV-1896][crit:High][vendor:cnv-qe@redhat.com][level:component] Should not allow cloning into a smaller sized data volume", func() {

File: tests/cloner_test.go [hunk 6]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+		Expect(err).ToNot(HaveOccurred())
+		Expect(f.VerifyTargetPVCContentMD5(f.Namespace, utils.PersistentVolumeClaimFromDataVolume(targetDv2), matchFile, md5sum[:32])).To(BeTrue())
+		_, err = utils.WaitPodDeleted(f.K8sClient, "verify-pvc-md5", f.Namespace.Name, verifyPodDeletedTimeout)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Waiting for clone to be completed")
+		err = utils.WaitForDataVolumePhase(f.CdiClient, f.Namespace.Name, cdiv1.Succeeded, targetDv3.Name)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(f.VerifyTargetPVCContentMD5(f.Namespace, utils.PersistentVolumeClaimFromDataVolume(targetDv3), matchFile, md5sum[:32])).To(BeTrue())
+	})
```
[the remark this answers is not in the document]
Comment (awels) [reply tests/cloner_test.go]: Done

File: tests/cloner_test.go (hunk 4 above)
Comment (awels) [reply tests/cloner_test.go]: Done

File: tests/cloner_test.go [hunk 7]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+		if targetDv2 != nil {
+			By("Cleaning up target2 DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv2.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+		if targetDv3 != nil {
+			By("Cleaning up target3 DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv3.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
```
[the remark this answers is not in the document]
Comment (awels) [reply tests/cloner_test.go]: Yes, done

File: tests/cloner_test.go (hunk 5 above)
Comment (awels) [reply tests/cloner_test.go]: Done

File: tests/cloner_test.go [hunk 8]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+
+		By("Cloning from the target1 DataVolume to target3")
+		targetDv3 = utils.NewDataVolumeForImageCloning("target-dv3", "500Mi", f.Namespace.Name, targetDv1.Name)
+		_, err = utils.CreateDataVolumeFromDefinition(f.CdiClient, f.Namespace.Name, targetDv3)
+		Expect(err).ToNot(HaveOccurred())
+		By("Waiting for clone to be completed")
+		err = utils.WaitForDataVolumePhase(f.CdiClient, f.Namespace.Name, cdiv1.Succeeded, targetDv3.Name)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(f.VerifyTargetPVCContentMD5(f.Namespace, utils.PersistentVolumeClaimFromDataVolume(targetDv3), matchFile, md5sum[:32])).To(BeTrue())
+	})
```
[the remark this answers is not in the document]
Comment (awels) [reply tests/cloner_test.go]: Done

File: tests/cloner_test.go (hunk 1 above)
Comment (awels) [reply tests/cloner_test.go]: Done

File: tests/cloner_test.go [hunk 9]
```diff
@@ -75,6 +76,219 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+			By("Cleaning up target2 DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv2.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+		if targetDv3 != nil {
+			By("Cleaning up target3 DV")
+			err = utils.DeleteDataVolume(f.CdiClient, f.Namespace.Name, targetDv3.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+	})
```
[the remark this answers is not in the document]
Comment (awels) [reply tests/cloner_test.go]: Done

**Comment by awels** [comment]
ci test please

File: tests/cloner_test.go [hunk 10]
```diff
@@ -75,6 +76,166 @@ var _ = Describe("[rfe_id:1277][crit:high][vendor:cnv-qe@redhat.com][level:compo
...
+		// By not waiting for completion, we will start 3 transfers in parallell
+		By("Cloning from the source DataVolume to target1")
+		targetDv1 = utils.NewDataVolumeForImageCloning("target-dv1", "500Mi", f.Namespace.Name, sourceDv.Name)
+		_, err = utils.CreateDataVolumeFromDefinition(f.CdiClient, f.Namespace.Name, targetDv1)
+		Expect(err).ToNot(HaveOccurred())
+		By("Cloning from the source DataVolume to target2 in parallel")
+		targetDv2 = utils.NewDataVolumeForImageCloning("target-dv2", "500Mi", f.Namespace.Name, sourceDv.Name)
+		_, err = utils.CreateDataVolumeFromDefinition(f.CdiClient, f.Namespace.Name, targetDv2)
+		Expect(err).ToNot(HaveOccurred())
+		By("Cloning from the target1 DataVolume to target3 in parallel")
```
Comment (aglitke) [inline tests/cloner_test.go]: Is it possible that this clone would start before the first one (source -> target1)?  In that case you'd have an empty target1 volume being cloned to target3

Comment (awels) [reply tests/cloner_test.go]: Yes, that is a good source of heisenbugs. Let me change that to just be the source.

**Comment by awels** [comment]
/hold cancel
Fixed the failing tests.

**Pull request merged**